### PR TITLE
feat(server): Store information about available checkers to the database

### PR DESCRIPTION
**Pull request by whisperity**
This patch contains the back-end developments and solutions to collecting the information needed for #4049.

Summary
-------

In #4049, it was deemed important that losing the information about which checkers executed during analysis is confusing to developers who have access only to the web-based interface without anything from the `analyze` command (such as CI logs). The confusion stems from checkers that were _enabled_ but did not produce _any_ reports not showing up anywhere on the interface. This is exasperated by the fact that the "analysis command" that is currently shown on the Web UI often contains references to random temporary files created by CI systems without seeing their contents. 

The goal of this initiative is to deterministically show the knowledge gathered from the `metadata.json`s (if applicable) about which analysers' which checkers executed. Having this list at hand is useful from an auditing point of view, as well, to ensure that it can be seen whether the executed analysers contained a particular checker or not. Thus, we move from the following distinction:

* _"checker that produced a report"_
* _"checker that did not produce a report"_

over to a more descriptive one:

* _"checker that was disabled"_
* _"checker that was enabled" (and produced a report)_
* _"checker that was enabled" (but did not produce a report)_

Description
-----------

This patch implements the **back-end** changes required to support this new feature. No _front-end_ changes available, yet.

1. First, a **new** table, `checkers` is created, which deterministically stores a unique ID (per database) for each `(analyser_name, checker_name)` pair. The uniqueness of IDs to, henceforth, checkers is enforced on the database level with a compound `UNIQUE KEY`.
2. `checkers` is filled with the names of all checkers identified during a run's store in a transactionally separated way, preceding the actual _Report_ storage process. This ensures that during a (potentially concurrent) storage process, constraint violations do not tear down the progress of an entire run's contents. (A generously timed retry logic is implemented to ease this process, but it is not guaranteed to be infallible.)
3. `analysis_info`, which previously stored only the _`analyze` command-line_ is extended to `JOIN` onto the `checkers` table and stores a single `BOOL` to say whether the joined checker was enabled in the joined `analysis_info`.
4. As there is a unique `analysis_info` for all `run_histories`, this list will trivially work in the same way as retrieving the _`analyze` command-line_ for a past snapshot of a _Run_ currently work.
    * During incremental schema migration, this information is **NOT** retroactively conjured. This is because there is currently no way of telling which checkers were **"enabled"** in an analysis, as potentially _deleted_ reports of existing `Run`s might skew these results. It's safer (or at least less misleading) just to say _`"No information."`_ for runs that were stored prior to the upgrade.
5. The _AnalysisInfo_ type on the Thrift API is extended to contain the information (`map<analyser_name: string, map<checker_name: string, enabled: bool>>`) now stored in the database as of point _3._. 
6. Several checker-related aspects that were previously stored for every _Report_ instance individually, most importantly the **Severity**, are moved into `checkers` instead. This simplifies the storage and server-start clean-up logic somewhat while complicating matters elsewhere during querying. The following columns are removed (essentially, uniqued out) from _Reports_: `checker_id` (string, "name"), `analyzer_name`, `severity`.
    * During incremental schema migration, the data from `reports` is pre-collected and dumped to `checkers`, with the new `FOREIGN KEY`s appropriately established as well. This means that for a _product_ with a diverse set of checkers that all produced warnings, the vast majority of the _"potential checker ID universe"_ will practically be allocated during the schema upgrade.
    * During schema downgrades, there is a clean way of doing this in the opposite direction, and the contents of these _Report_ columns are restored appropriately.
7. The `checker_cat`(-egory, string) and `bug_type` (string) columns of _Reports_ is dropped **irrevocably**, and not restored during a downgrade. It was never clear whether these values were uniquely the same for a particular checker, so moving these to the `checkers` table might not be a feasible solution. However, as these columns contained generic data that was never exposed over the API, the storage of them is deemed unnecessary. The semantics for whatever was available in these columns were never actually defined and were very analyser-specific or specific to how the _report-converter_ dealt with the raw output of an analyser.

Miscellaneous
-------------

This patch also contains the following **quality improvements** that are not directly related to the feature implementation but were designed and developed in parallel to this feature's development.

* This patch changed some of the code related to how a *product* database is connected during schema migration. There were two connections here, one for the *product* database (under migration), and around that, to the *config* database (to get the connection string for the product). This is to ensure that a long migration (such as the one present due to this patch) will not cause nasty time-outs on the *config* connection. If that happens, the resulting exception (of no meaningful value, because the *config* database is only **read**, and only **once** per product under migration) previously escaped and killed the entire server, which is not nice for potential overnight migrations. The new version guards that such exceptions will not escape, the server stays alive, and even if the migration of one product fails, the subsequent product migrations are still attempted by the same server execution, so there is no need for any restarts or auto-restart services.
* Change from the deprecated `Binary` type to `LargeBinary` consistently over the database schema. This means that this patch supersedes #3736. To answer @vodorok's [question added to #3736](http://github.com/Ericsson/codechecker/pull/3736#pullrequestreview-1301878104), [SQLAlchemy's source code](http://github.com/sqlalchemy/sqlalchemy/blob/rel_1_3_23/lib/sqlalchemy/sql/sqltypes.py#L987-L994) shows the established inheritance relationship: `_Binary` ${<:}$ `LargeBinary` ${<:}$ `Binary`, and `Binary.__init__()` only delegates `LargeBinary.__init__()`. Thus, it is clear that the two types behave the exact same across all database dialects, and the changes of both #3736 and this patch are safe without additional schema migration commands.
* Added the possibility of **logging** output during schema migrations. This was necessitated by the complexity of the steps of the migration script `c3dad71f8e6b_...` of this patch. We will be updating multiple millions of rows in chunks, in multiple passes (one read, one collate, one write), so the accountability of where the migration is going benefits a lot from these improvements.
* The in-memory data structures that represent parsed `metadata.json`s previously stored the `checkers` "list" in a hard-to-understand and unwieldy-to-use data structure: `checkers: Dict[checker_name: str, _: Union[enabled_checkers: List[name: str] | _: Dict[name: str, enabled: bool]]]`. This made accessing actual data from this troublesome, as every client code (including new things I was writing for this patch) had to branch on whether the actual available data was "old-style" (only the list of **enabled** checkers) or "new-style" (the list of all found checkers, marked with its enabledness bit). What's more, the tests were conflating what it means to be a `v1` `metadata.json`. The original implementation, up to 7254d05a8b7262e4979ac613f32d6c3e0aa0d3cc in **September 2017**, contained the former `List[str]` grouped by the analyser. In **March 2020**, bd775d60950d48884b8f1dc83d8b82653b83cfa3 introduced `v2` files, but at this point, `v1` was already doing `Dict[str, bool]`s. In between the two points, in **December 2019**, 0cd28acac7d31e4a0260c147b8e803b7a36908f0 changed the format without the concept of "format versions", turning the list into `Dict[str, bool]`. Nevertheless, several test files that do not indicate the version still test whether the "real" `v1` is understood by the system. I fixed all these issues by transitioning to the current _`v2`_ format **in memory** as soon as possible, creating an appropriate `checkers: Dict[str, Dict[str, bool]]` data structure irrespective of the format of the `metadata.json` in the input.

### CODE REVIEW

**Review by vodorok — CHANGES_REQUESTED** [review]
During database migration, the server gets trapped in a cyclic restart with the following error:
```
Traceback (most recent call last):
  File "/codechecker/lib/python3/codechecker_common/cli.py", line 209, in main
    sys.exit(args.func(args))
  File "/codechecker/lib/python3/codechecker_server/cmd/server.py", line 425, in __handle
    main(args)
  File "/codechecker/lib/python3/codechecker_server/cmd/server.py", line 1009, in main
    server_init_start(args)
  File "/codechecker/lib/python3/codechecker_server/cmd/server.py", line 936, in server_init_start
    __db_migration(cfg_sql_server, context.run_migration_root,
  File "/codechecker/lib/python3/codechecker_server/cmd/server.py", line 609, in __db_migration
    ret = db.upgrade()
  File "/codechecker/lib/python3/codechecker_server/database/database.py", line 326, in upgrade
    command.upgrade(cfg, "head")
  File "/usr/local/lib/python3.9/site-packages/alembic/command.py", line 294, in upgrade
    script.run_env()
  File "/usr/local/lib/python3.9/site-packages/alembic/script/base.py", line 490, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/usr/local/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 97, in load_python_file
    module = load_module_py(module_id, path)
  File "/usr/local/lib/python3.9/site-packages/alembic/util/compat.py", line 182, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/codechecker/lib/python3/codechecker_server/migrations/report/env.py", line 120, in <module>
    run_migrations_online()
  File "/codechecker/lib/python3/codechecker_server/migrations/report/env.py", line 114, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/usr/local/lib/python3.9/site-packages/alembic/runtime/environment.py", line 813, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/usr/local/lib/python3.9/site-packages/alembic/runtime/migration.py", line 560, in run_migrations
    step.migration_fn(**kw)
  File "/codechecker/lib/python3/codechecker_server/migrations/report/versions/c3dad71f8e6b_store_information_about_enabled_and_disabled_checkers_for_a_run.py", line 235, in upgrade
    upgrade_analysis_info()
  File "/codechecker/lib/python3/codechecker_server/migrations/report/versions/c3dad71f8e6b_store_information_about_enabled_and_disabled_checkers_for_a_run.py", line 62, in upgrade_analysis_info
    _, new_analyzer_command = recompress_zlib_as_tagged_exact_ratio(
  File "/codechecker/lib/python3/codechecker_server/migrations/common.py", line 89, in recompress_zlib_as_tagged_exact_ratio
    data = raw_zlib_decode_buf(value)
  File "/codechecker/lib/python3/codechecker_server/migrations/common.py", line 26, in raw_zlib_decode_buf
    return zlib.decompress(value)
TypeError: a bytes-like object is required, not 'NoneType'

```

**Comment by whisperity** [comment]
> During database migration, the server gets trapped in a cyclic restart with the following error:

@vodorok I have identified the cause of the problem for that behaviour and fixed it. Turns out I made the bogus assumption that `analysis_info.analyzer_command` would be `NON NULL` in every case. The fix was trivial (`if ... is None: continue` :smirk:), I will update the patch soon.

Unfortunately, I encountered some pretty weird behaviour that I need to track down before this patch is touched...

File: web/server/codechecker_server/database/db_cleanup.py [hunk 1]
```diff
@@ -223,13 +172,70 @@ def remove_unused_analysis_info(session_maker):
...
+                           Checker.severity) \
+                    .filter(Checker.analyzer_name == analyzer) \
+                    .all()
+                for checker_row in checkers_for_analyzer_in_database:
+                    checker: str = checker_row.checker_name
+                    old_severity_db: int = checker_row.severity
+                    old_severity: str = \
+                        Severity._VALUES_TO_NAMES[old_severity_db]
+                    new_severity: str = \
+                        checker_labels.severity(checker, analyzer)
```
Comment (whisperity) [inline web/server/codechecker_server/database/db_cleanup.py:216]: There might be a slight issue here, because `(analyzer, checker)` comes from the database, which means it will be, for example `(cppcheck, unusedVariable)`,  but the config file, unfortunately, contains `cppcheck-unusedVariable` as the _checker name_, resulting in the severity becoming wiped.

I will do two things to fix this:

 1. Of course, fix the logic here, so the "new style" checker names are also attempted from the config file.
 2. Prevent the wiping of severity to `0` in addition to giving an explicit :warning: `[WARNING]` about the fact.

**Review by whisperity — COMMENTED** [review]
I think the real data testing surfaced a few issues with which I'm converging to a solution...

**Review by bruntib — CHANGES_REQUESTED** [review]
I submit some relatively minor comments I had so far. I'm still reviewing this ticket, so maybe some further comments arrive later. By the way, it's a very nice job, great improvement in CodeChecker!

File: web/server/codechecker_server/metadata.py [hunk 2]
```diff
@@ -61,32 +66,29 @@ def __init__(self, metadata_file_path):
...
-                    self.checker_to_analyzer[checker_name] = analyzer_name
-                    if enabled:
-                        self.enabled_checkers.add(checker_name)
-                    else:
-                        self.disabled_checkers.add(checker_name)
-            else:
-                self.enabled_checkers.update(analyzer_checkers)
-
-                for checker_name in analyzer_checkers:
-                    self.checker_to_analyzer[checker_name] = analyzer_name
```
[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/metadata.py:89]: As discussed in the original post of the patch:

> I fixed all these issues by transitioning to the current v2 format in memory **as soon as possible**, creating an appropriate `checkers: Dict[str, Dict[str, bool]]` data structure irrespective of the format of the `metadata.json` in the input.

Support is not lost (as observed from the tests still existing for such old-school inputs, and those tests passing). But every function that deals with querying the loaded structure now uses a unified format. Conversion to this format is done when loading the data.

File: web/tests/functional/report_viewer_api/test_run_data.py [hunk 3]
```diff
@@ -171,3 +174,67 @@ def test_sort_run_data(self):
...
+
+        checkers = info.checkers
+
+        def assertChecker(analyzer, checker):
+            self.assertTrue(checkers[analyzer][checker].enabled)
+
+        def assertNotChecker(analyzer, checker):
+            self.assertFalse(checkers[analyzer][checker].enabled)
+
+        assertNotChecker("clangsa", "alpha.cplusplus.MismatchedIterator")
```
[the remark this answers is not in the document]
Comment (whisperity) [reply web/tests/functional/report_viewer_api/test_run_data.py:220]: What should I do in this case? Do we have a well-known earliest version that is supposed to be supported? Otherwise, #4116. It's incredibly stupid that running these tests requires you to have an analysis environment...

File: web/server/codechecker_server/metadata.py [hunk 4]
```diff
@@ -61,32 +66,29 @@ def __init__(self, metadata_file_path):
         self.disabled_checkers: DisabledCheckers = set()
         self.checker_to_analyzer: CheckerToAnalyzer = dict()
 
-        self.__metadata_dict = {}
+        self.__metadata_dict: Dict[str, Any] = {}
         if os.path.isfile(metadata_file_path):
-            self.__metadata_dict = load_json(metadata_file_path, {})
+            self.__metadata_dict = cast(Dict[str, Any],
+                                        load_json(metadata_file_path, {}))
```
[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/metadata.py:72]: No, `Dict` is `Dict[Any, Any]`. Or, if we want to be pedantic, it's `Dict[Hashable[Any], Any]`, but, at least with CPython, everything is hashable, worst case you just trivially hash to pointer identity.

```py3
$ python3
Python 3.9.2 (default, Feb 28 2021, 17:03:44)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> ds = dict()
>>> a = object()
>>> b = 1
>>> ds[a] = "Foo"
>>> ds[b] = a
>>> from pprint import pprint
>>> pprint(ds)
{1: <object object at 0x7f36ddcff1c0>, <object object at 0x7f36ddcff1c0>: 'Foo'}
```

Unfortunately, this is not always deducible. A top-level JSON object does **not** have to be a dict. A top-level JSON can be a `List` (e.g., `compile_commands.json` is essentially a `List[Dict[str, ?]]`), or some scalar.

File: web/server/codechecker_server/api/mass_store_run.py [hunk 5]
```diff
@@ -191,6 +229,14 @@ def get_blame_file_data(
     return blame_info, remote_url, tracking_branch
 
 
+def checker_name_for_report(report: Report) -> Tuple[str, str]:
+    analyzer_name, checker_name = \
+        getattr(report, "analyzer_name"), getattr(report, "checker_name")
+    analyzer_name = analyzer_name if analyzer_name else "UNKNOWN"
+    checker_name = checker_name if checker_name else "NOT FOUND"
```
[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/api/mass_store_run.py:236]: Raw plists not generated to the exact format we usually consume: https://github.com/Ericsson/codechecker/blob/v6.23.0/web/tests/functional/cppcheck/test_proj/divide_zero.plist

Also handled existing behaviour:

https://github.com/Ericsson/codechecker/blob/v6.23.0/web/server/codechecker_server/api/report_server.py#L2593-L2598

https://github.com/Ericsson/codechecker/blob/v6.23.0/web/server/codechecker_server/api/mass_store_run.py#L732

File: web/server/codechecker_server/api/mass_store_run.py [hunk 6]
```diff
@@ -529,6 +575,66 @@ def __add_file_content(
...
+        eagerly populated, otherwise the Python bindings will fail.
+        However, eager population will result in exceptions that a flush was
+        created before the transaction was complete.
+
+        Moreover, this is performed separately from the storing of the details
+        of a run to reduce contention if two parallel stores, especially across
+        server instances (in a distributed/load-balanced environment) want to
+        store the same identifier(s).
+        """
+        max_tries, tries, wait_time = 3, 0, timedelta(seconds=30)
```
[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/api/mass_store_run.py:595]: `(analyzer_name, checker_name)` has a `UNIQUE` constraint on them. Every conceptually existing checker will need to have at most **one** `checker_id` for the look-up to work. If two `store`s happen concurrently and they store the same set of checker names (highly likely), the storage that ends up `COMMIT`ting later will explode due to the constraint violation. This is an existing issue with everything that is supposedly stored as a "product-global" information, e.g., [files and contents](https://github.com/Ericsson/codechecker/blob/v6.23.0/web/server/codechecker_server/api/mass_store_run.py#L132), [review statuses](https://github.com/Ericsson/codechecker/blob/v6.23.0/web/server/codechecker_server/api/mass_store_run.py#L1238-L1259).

In fact, in this patch, I removed the above second logic, due to having previously moved review statuses into individual `report` rows. But the general issue with the logic now comes up again with the introduction of a new product-global information.

```py3
# TODO: Since review status is stored in "reports" table and
# "review_statuses" table is not written during storage, this
# multiple trials should be unnecessary.
```

[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/api/mass_store_run.py:595]: The only downside of, essentially pre-committing checker names, is that **if** a store introduces a new checker name, let's say `foo/bar`, and the report storage fails, only the report storage is rolled back, because the ID for the checkers are already there and committed. So, we will store a little bit of _potentially_ unused data in the database. But it's two generally small strings with a number, so like... guesstimateable to be at most 256 bytes-ish.

File: web/server/codechecker_server/api/mass_store_run.py [hunk 7]
```diff
@@ -191,6 +229,14 @@ def get_blame_file_data(
     return blame_info, remote_url, tracking_branch
 
 
+def checker_name_for_report(report: Report) -> Tuple[str, str]:
+    analyzer_name, checker_name = \
+        getattr(report, "analyzer_name"), getattr(report, "checker_name")
+    analyzer_name = analyzer_name if analyzer_name else "UNKNOWN"
+    checker_name = checker_name if checker_name else "NOT FOUND"
+    return (analyzer_name, checker_name)
```
[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/api/mass_store_run.py:237]: Unfortunately, the new version is **not** equivalent to what is in the patch. (Note that type **hints** are just hints, there is no guarantee that indeed a conforming `Report` object is passed to this function!) I don't have clear memories of this anymore due to having worked multiple months on this feature, but I would wager that the [mentioned raw Cppcheck plist test file](https://github.com/Ericsson/codechecker/blob/v6.23.0/web/tests/functional/cppcheck/test_proj/divide_zero.plist) was the way to trigger the inequivalence.

The coalescing `x or y` will only return `y` if `x` is successfully evaluated, and returns falsy. A direct `.` operator, however, may result in an `AttributeError`, something that is only coalesced by `getattr()`'s implicit `default = None` parameter.

```py3
$ python3
Python 3.9.2 (default, Feb 28 2021, 17:03:44)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class C:
...   def __init__(self):
...     self.member = "X"
...     self.member2 = None
...
>>> c = C()
>>> (c.member or "default 1", c.member2 or "default 2")
('X', 'default 2')
>>>
>>>
>>>
>>> a = object()
>>> (a.member or "default 1", a.member2 or "default 2")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'object' object has no attribute 'member'
```

File: web/server/codechecker_server/api/mass_store_run.py [hunk 8]
```diff
@@ -61,10 +65,44 @@ def __enter__(self, *args):
...
+        return self
+
+    def __exit__(self, *args):
+        self.__run_lock = None
+        self.__session = None
+
+    def noop(self):
+        """
+        Does nothing, but prevents Pylint from complaining, and prevents the
+        context from being optimised out.
```
[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/api/mass_store_run.py:101]: Unused variable, unused parameter... Note that I moved the code over to the context manager here. It would be weird if the `with:` statement let optimise out one of the contexts.

As suspicious as it is, the existing code already contains provisions against some optimisations, here:

https://github.com/Ericsson/codechecker/pull/4089/files#diff-857761db3a331e1e83bd931edaf83b01725c9b2c8d755a3d6257334ffad17d4b

I was not able to reproduce, neither pro nor contra, what this instruction was trying to prevent, and the blame history ran dry quickly as well, so I opted to keep the existing behaviour intact, even if hidden behind a small refactoring.

File: web/api/report_server.thrift [hunk 9]
```diff
@@ -467,8 +468,16 @@ union AnalysisInfoFilter {
   3: i64 reportId,
 }
 
+struct AnalysisInfoChecker {
+  1: optional bool enabled,  // If the checker was enabled during the analysis.
```
[the remark this answers is not in the document]
Comment (whisperity) [reply web/api/report_server.thrift:472]: I adhered to the [Thrift guidelines](https://diwakergupta.github.io/thrift-missing-guide/#defining-structs). If you don't say something is `optional`, it will be impossible to accurately deprecate and drop without breaking something. (The API file is full of `!DEPRECATED!` markers already because we did not care about this for a very long time.) In general, new API functions seemed to be often introduced as taking a `struct` member instead of multiple parameters. In returns values, changing from `map<X, Y>` to `map<X, struct<Y, Z>>` would be a horrible ABI break.

At a later point in time, maybe someone will decide that the enabled-ness of a checker is not enough, they want to obtain the exact checker options? That could be a reasonable addition here...

Or the notion of `enabled`-ness can change from a `bool` to something else, because we realise things could need to be transitively enabled...

Having this `optional` allows handling the lack of the value in the client codes without Thrift blowing up under us.

[the remark this answers is not in the document]
Comment (bruntib) [reply web/api/report_server.thrift:472]: I'm not sure if I agree with the first part. It would imply that every struct member and function parameter should be automatically optional. I prefer thinking of "optional" as a non-required thing, and not as a tool for providing backward compatibility. As far as I understand (please, correct me if I'm wrong) this optional provides backward compatibility during a potential API change by not requesting this field anymore. But if we decide to deprecate it, then we still need to leave this in the source code with a `!DEPRECATED!` label for the sake of old clients. Something makes me think that the number at the beginning of the line is used for supporting API changes.
My biggest technical concern about optional fields is that we are messaging to the clients that this field is really not necessary, and we need to check its existence at every single place where we use it. I'm not convinced it would be convenient, regarding that it really is a required field.

File: web/server/codechecker_server/api/mass_store_run.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bruntib) [reply web/server/codechecker_server/api/mass_store_run.py:101]: Sorry, I don't really understand the situation. Unfortunately the link doesn't open for me properly, because GitHub collapses the files with too many changes. So could you, please, elaborate a little more? Do we need this noop if it doesn't reproduce the issue, then? :smile:

File: web/server/codechecker_server/api/mass_store_run.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (bruntib) [reply web/server/codechecker_server/api/mass_store_run.py:237]: I would say that we should rely on type-hints, that's why we use them. If we expect a `Report` object, then we may trust it has the required fields. This is the exact situation what I meant at a previous comment: if we are distrustful with the caller, it complicates a lot at the callee side. So, if this function gets something else than a `Report`, then we should fix that call. Another option is to extend the type-hint to indicate other possible parameter types too, but I would prefer the first option.

File: web/server/codechecker_server/api/mass_store_run.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bruntib) [reply web/server/codechecker_server/api/mass_store_run.py:236]: See my previous comments.

File: web/server/codechecker_server/metadata.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (bruntib) [reply web/server/codechecker_server/metadata.py:72]: Oh, I mixed it with JS :smile: Thanks for the info!

File: web/server/codechecker_server/api/mass_store_run.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/api/mass_store_run.py:101]: @bruntib Here is the mentioned logic in the release version: https://github.com/Ericsson/codechecker/blob/v6.23.0/web/server/codechecker_server/api/mass_store_run.py#L1259-L1276

I would prefer to keep this noop, because I can neither prove nor disprove that removing it would result in the issue not reappearing. It seems that this behaviour is tied to some specifics of concurrent execution over multiple servers, something that is not simple enough to reproduce locally.

It seems it was I who introduced this in 17ddab02615fec2dd24733a2df9e76cb0a953f9f, but it was 7 years ago, and I do not remember contexts that far back in time.

**Review by vodorok — DISMISSED** [review]
Please add this to a separate PR

File: web/server/codechecker_server/api/mass_store_run.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/api/mass_store_run.py:237]: > we should fix that call

It's a call through the API, which involves what is being put into the mass-store ZIP. **If** the bogus behaviour here (fields being unknown) happen because of plists NOT coming from the report-converter are being uploaded, then this means that a full-scale schema validation logic must be developed into the **`store`** (and the _`parse`_?) command such that it doesn't allow upgrading.

The exact details of what format of `plist`s is accepted over the API is not documented anywhere, so we might be breaking a lot of downstream projects (VSCode Plugin? Eclipse Plugin? Third-party users who create their own `plist`s from analyzers not passed through the report converter?) as well.

Or should we just let the `AttributeError` escape here and explode the entire storage of the user, and let the user figure out why their reports are broken?

File: web/server/codechecker_server/api/mass_store_run.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/api/mass_store_run.py:236]: (Side note: `previous` here is undefined, because the comments are ordered, I think, by the hunk they are attached to, irrespective of when they were typed in.)

File: web/api/report_server.thrift (hunk 9 above)
[the remark this answers is not in the document]
Comment (whisperity) [reply web/api/report_server.thrift:472]: > I prefer thinking of "optional" as a non-required thing, and not as a tool for providing backward compatibility.

The guidebook recommended by Thrift developers disagrees with this (emphasis mine):

> ⛔ **Required Is Forever**
> 
> You should be very careful about marking fields as required. If at some point you wish to stop writing or sending a required field, it will be problematic to change the field to an optional field — **old readers will consider messages without this field to be incomplete and may reject or drop them unintentionally.**

I based the changes on existing observed practice, e.g., https://github.com/Ericsson/codechecker/commit/4b4d2878f8fef57b5b334083470ee0034e825451, https://github.com/Ericsson/codechecker/commit/7b0dc2ede9e9ef214875f869260de55621362493 added new fields that are `optional`.

In https://diwakergupta.github.io/thrift-missing-guide/#_versioning_compatibility, it says that

> Any new fields that you add should be optional.

(Okay, technically, this patch here introduces the entire struct (all its one member), so this isn't just a "new field", but a whole new type...)

You are right that the number is used as the "slot" which Thrift internally uses to identify which piece of data belongs where. The *name* of the fields isn't serialised. This helps reduce the size of the transmitted packet. This is not directly related to API changes.

> But if we decide to deprecate it, then we still need to leave this in the source code with a `!DEPRECATED!` label for the sake of old clients.

This is only true if you are still sending that information. However, in many cases, we actually do not. For example, the run's deprecated `checkCommand` became `analysisInfo`, and when queried, an empty string (albeit a valid result...) is returned.

The problem is that with `bool` there isn't a good third option that could be the "default value once the field is deprecated", because both `false` and `true` mean something very specific in this context.

> Non-required fields can be removed, as long as the tag number is not used again in your updated message type (it may be better to rename the field instead, perhaps adding the prefix "OBSOLETE_", so that future users of your .thrift can’t accidentally reuse the number).

File: web/server/codechecker_server/database/common.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+    If 'kind' is "str", the decoded payload will be decoded into a string.
+    Otherwise, it is returned as a raw 'bytes' buffer.
+    """
+    buffer, kind, _ = from_zlib_tag_prefix(buffer)
+    buffer = zlib.decompress(buffer)
+    return kind, buffer.decode() if kind == "str" else buffer
+
+
+class ZLibCompressedString(TypeDecorator):
```
[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/database/common.py:83]: Sure! It has grown vestigial in the purview of this patch anyways, considering we're no longer going to store the `metadata.json` in the DB as a single entity.

File: web/server/codechecker_server/api/mass_store_run.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/api/mass_store_run.py:236]: Also, the lack of a checker name being set to `"NOT FOUND"` was here: https://github.com/Ericsson/codechecker/blob/v6.23.0/web/server/codechecker_server/api/mass_store_run.py#L750

Now, because the string is changed to an ID that participates in a `FOREIGN KEY` lookup, and especially because it is an eagerly loaded sub-object, we have to explicitly obtain a **valid** ID, even for the unknown cases.

Unfortunately, it could also be set to **empty string** if you end up generating a `plist` which explicitly defines the corresponding key to be empty, instead of being *undefined*. We have evidence of people using CodeChecker with their own custom plist-generating infrastructure (or custom tools that somehow generate plists in the end), so saying that we only support whatever the report-converter is capable of spitting out, would also break these use cases. In the report-converter, the **checker name** is coalesced to unknown (https://github.com/Ericsson/codechecker/blob/v6.23.0/tools/report-converter/codechecker_report_converter/report/parser/plist.py#L246), but the **analyser name** is not. That was NULL-coalesced on the database level by setting `"unknown"` be the default value for the column in the schema.

```xml
<key>check_name</key><string></string>
```

[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/api/mass_store_run.py:236]: Nevertheless, I checked all the edge cases I could find and figured out a way to simplify the related logic in multiple places.

File: web/server/codechecker_server/api/mass_store_run.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (whisperity) [reply web/server/codechecker_server/api/mass_store_run.py:237]: Nevertheless, I checked all the edge cases I could find and figured out a way to simplify the related logic in multiple places.

File: web/tests/functional/report_viewer_api/test_run_data.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (whisperity) [reply web/tests/functional/report_viewer_api/test_run_data.py:220]: The `MismatchedIteratorChecker` has been part of LLVM since version **`8.0`**. This was the earliest commit I could identify where it was introduced, and has not been moved between packages ever since: https://github.com/llvm/llvm-project/commit/21583b733af14997718ab5aaf2a926821d479dd7#diff-e1b5fd68572f20de546330e9ed6537c1276f09f78f7ba7aac45b4fd887b7cc7bR1491

[the remark this answers is not in the document]
Comment (whisperity) [reply web/tests/functional/report_viewer_api/test_run_data.py]: I removed tests for `alpha.`, and instead opted for checking that a `debug.` checker is indeed disabled, as it almost certainly always is. 🙂

**Review by bruntib — APPROVED** [review]

File: web/server/codechecker_server/cmd/server.py [hunk 11]
```diff
@@ -499,15 +508,20 @@ def check_product_db_status(cfg_sql_server, migration_root, environ):
...
-    sess.close()
-    engine.dispose()
+        try:
+            status = db.connect()
+            s_ver = db.get_schema_version()
+            if s_ver in db_errors:
+                s_ver = None
+            prod_status[pd.endpoint] = (status, s_ver, package_schema,
+                                        db_location)
+        except Exception:
```
Comment (vodorok) [inline web/server/codechecker_server/cmd/server.py:519]: Was there a concrete exception occuring here? Shouldn't we expect for that?

**Review by vodorok — CHANGES_REQUESTED** [review]
Please do a measurement the store time of Xerxes-C analyzed with --enable-all. If the store time hit is not severe (under 5%), the patch can be merged. My other comments are questions, or suggestions.

File: web/server/codechecker_server/cmd/server.py (hunk 11 above)
Comment (whisperity) [reply web/server/codechecker_server/cmd/server.py:519]: The main exception here was from the absolutely unnecessary `sess.commit()` **timing out**. After removing that line from here and moving it before the products are iterated, I was not able to produce an exception here. However, the entire server, in principle, should NOT die just because one product had an issue, so I think a catch-all case is appropriate here. Keeping the server alive should be paramount.

**Comment by whisperity** [comment]
> Please do a measurement of the store time of Xerxes-C analyzed with `--enable-all`.

@vodorok Thanks for the mention of the potential issue. Indeed, there were problems with the added `flush()` call. Here are some comparative measurements, with the original report directory created with a 6.23 *release* at 2.8 GiB. Without this patch, the store (as logged by the timing of `massStoreRun()`) took **1140.93 s**, when using `--verbose debug` on both server and clientside.

With this patch, this had gone up to **1731.29 s**. However, undoing the `flush()` calls and changing the logic around it slightly to remove the original need they were introduced for only lessened this to **1644.56 s**.

So I've gone around and changed a few more things, because it turns out doing a `SELECT id FROM checkers WHERE column1 = <string> AND column2 = <string>` is not a good idea... :yum: With the upcoming changes to the patch, we will be down to **956 s**. (At first this was indeed strange for me, so I reproduced it three more times each time flushing and re-filling (to almost the brink) the system RAM to evade/prevent OS-level caching and spawning new database containers as well), and this reproduces within _2-3 s_ of error.

> If the store time hit is not severe (under 5%), the patch can be merged.

Unfortunately, it looks like this is a **16.2%** reduction in time required on this test. I hope that's not a problem. :smirk:

**Review by vodorok — APPROVED** [review]
Well, in this case, I think I can make an exception and allow a deviation from the 5% run time change  :).  Great work! 
LGTM!
